### DISCUSSION TITLE
Eigenvector space and eRPA energy correction

### DISCUSSION THREAD

**Discussion by gabrielasd** [gabrielasd/pyeom#56] [General]
I was looking again at the energies evaluated from the reconstructed 2RDMs, checking if they changed or not after removing the zero eigenvalues due to ph-RPA right-hand side. 

The image below has a table with the energies (E_2rdm_phrpa) resulting from 5 test cases where I changed which eigenvectors were used to evaluate the TRDMs (I only used the current "asymmetric" solver in the base EOM module for this data)
![ph_erpa_trdms_vectorspace](https://user-images.githubusercontent.com/26312264/119056808-4e510700-b999-11eb-800a-808714a0013b.png)

My conclusion from these tests was that I could use the full eigenvector space of ph-RPA if I took care not to overcount the contribution of the TRDM to the 2RDM.
However, I haven't been able to figure out how to make a similar procedure work for the ph-EOM without a commutator on the right-hand side.

To give some more context, I had been picking which eigenvectors get used to evaluate the TRDMs selecting only those with eigenvalues higher than {0.01}^2.
Also, I was comparing the solutions using an eigensolver with the right-hand-side eigenvectors explicitly projected out (pruned solver from here on) with those from the current asymmetric eigensolver in the base EOM module (original solver).
One of the tests I did was to relax the constrain in the selected eigenvectors, picking all with eigenvalues >= zero, and because the solutions were the same irrespective of what solver I was using (pruned or original) I wondered why I had neglected the zero eigenvalues before. 

Does this makes sense to you @PaulWAyers @msricher

**Comment by PaulWAyers** [comment]
I'm a little confused. Can you explain what (a) (b) (c) (d) (e) are better? I'm not clear on what "cv" and "w" are.  I think the w is the eigenvalue cutoff? So (b) and (c) are saying that the negative or positive eigenspace is cut off?  

- I would not be surprised if the results depend on the cutoff but *not* whether you use the projection method or the generalized inverse. Those should both be the same or very similar I think. (I think it could be proved to be equal.) 
- I expect the results to be sensitive to the cutoff but if the cutoff is small, then only "hard zeros" are cut off, and one is insensitive to the cutoff value. 
- It would be surprising to me that cutting off the whole positive or negative spectrum (if that is what (b) and (c) are) would work, but it could, if the two sets of spectrum are decoupled as they could be on physical grounds, though it's not 100% obvious to me that it is true mathematically.
-  Due to Brillouin's theorem, I would have maybe expected the CIS case to be the same as HF, but that isn't happening. I'm not sure what (d) is and why it works correctly when combiend with (e) unless this is what's being talked about in terms of "overcounting."

**Reply by gabrielasd** [reply]
cv and w are the eigenvectors and eigenvalues, respectively, of ph-EOM (RPA) represented as: 
A cv = w S cv
In (a) and (b) only the positive side of the spectrum is taken (the cutt-off values were 0.01^2 and zero respectively); In (c) only the negative side, while in (d) and (e) the whole spectrum is taken.

1- Yes, I agree. What I was testing originally was that after using the projection method the results I had before haven't changed (otherwise I had implemented something wrong).
2- I see, that makes sense. I have one doubt related to the zero eigenvalues though. The equation to get the TRDMs depends on the eigenvectors and the one and two RDMs, but not explicitly on the eigenvalues. So for there to not be a change in the results (looking at (a) vs (b)) would mean that all eigenvectors from "hard-zero" eigenvalues are zero, or that the corresponding RDM terms are zero, right?
4- I must have some implementation problem in the CIS form of ph-EOM. I didn't include the CIS tests on the table above, I could open a separate issue or discussion for it. But basically, it gives the same as HF if I were to use the condition of the case (a), but not (b). The is something going on with the reconstruction of the 2-RDM in case (b) that makes it not meet the normalization condition. It also depends on what expression I use to evaluate the TRDM (no commutator or commutator).
The (d) case corresponds to using the eigenvectors of the full ph-RPA spectrum to evaluate the TRDMs. I think this should be what corresponds to the actual equation to approximate the TRDMs, as the sum runs over all excited states. But I may have a misinterpretation. And because
 cases (a) and (b), which used only one side of the spectrum, gave the same energies, I thought there was something being accounted for twice on (d). That was why for (e) I used a 1/2 factor on the TRDM matrix when reconstructing the 2RDM.

**Reply by PaulWAyers** [reply]
Ah, this is starting to make sense. .

2- A good point. I'm not sure. The "complete set of eigenvectors" includes eigenvectors with a zero eigenvalue, but if you write the RPA equation with explicit dependence on orbital energies these would be strange/singular cases. It requires detailed thought. 
4- Same as HF is sensible. In case b, you are screening based on "hard zeros?" It would be useful to look at the right-hand-side eigenvalues. It may be that 1e-2 or 1e-4 is too big as a screening, but 0.0 (as opposed to 1e-12 or something small but not zero) isn't, just because of round-off errors.
5- It makes sense that with the "double counting" you could get something wrong. Is the TRDM in case (a,b,c) exactly twice that of what happens in case (d), or is it just the energy? Are cases (a),(b),(c) all equal in terms of their TRDM?

**Reply by PaulWAyers** [reply]
- Try this for DOCI.
- We still need to look at the math to understand some things (point 2 above).

**Reply by gabrielasd** [reply]
For the DOCI reference wfn I observed the same behavior in the energies as for HF:
![dicussion_erpa_Ecorrection_doci](https://user-images.githubusercontent.com/26312264/122089353-ab8f7b00-cdd4-11eb-9e14-30ee5c3fcba3.png)

I also looked at the spectrum for ph-eRPA and there is the same symmetry? between the positive and negative sides of the spectrum as for HF.
And as happened for ph-RPA, the transition RDMs from case (a)/(b)/(c) and (e) are not equivalent, only the energies are (the multiplication by the two-electron integrals, in the 2nd term of the energy equation, is what makes the energies equal at the end)

I'm left with looking at the math and see if I can make sense of all these.

**Reply by PaulWAyers** [reply]
I think it makes sense that ph symmetry is always there, since ph is a fundamental symmetry. 

We should look at the math and see if we can make sense of it.

**Reply by gabrielasd** [reply]
The paragraph following Eq. 8 [here](https://pubs.acs.org/doi/10.1021/acs.jpclett.8b02391) suggests that the factor of two we have seen in the numerical tests above may be possible:
![relation_eigvects jpg](https://user-images.githubusercontent.com/26312264/122884492-36f49900-d30c-11eb-985e-42dbce260de2.jpg)

As a quick check, I looked at the RPA solutions for H2 STO-6G. For the $\omega_pos$ and $\omega_neg$ corresponding to the singlet transition, the relation is met, but for the triplet transitions (three pairs of energies) It was not evident to me if it was also the case.

So I thought that if the relation between the eigenvectors is as Pernal stated the terms summing over the positive and negative sides of the spectrum in the equation bellow should be equivalent:
![sum_tdms](https://user-images.githubusercontent.com/26312264/122885921-812a4a00-d30d-11eb-9249-7a41fdb0d7b7.jpg)
but that is not what I've seen from the numerical test so far.

Later I'll post (hopefully) better notes on the math

**Reply by PaulWAyers** [reply]
Seems like this is a good lead on what is happening.

**Comment by gabrielasd** [comment]
@PaulWAyers please see the following notes on the description of the problem treated in this discussion
[tdm_erpa_issue.pdf](https://github.com/gabrielasd/eomee/files/7187602/tdm_erpa_issue.pdf)

**Reply by PaulWAyers** [reply]
Thanks @gabrielasd . I think we really need to work through this in detail. I can't see anything obvious at first thought. I'll try to find another snippet of time tonight to get back into it, though.
